sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_tex_cache_read.asciidoc

```diff
@@ -0,0 +1,109 @@
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+This extension exposes the https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-ld-global-nc[ld-global-nc] ptx instruction so that users can load a register variable to the non-coherent read only texture cache. The texture cache is designed for random access reads and is most performant when work-items within a sub-group read a set of addresses of texture memory that are close to one another in the cache.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `sycl_ext_oneapi_cuda_tex_cache_read` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== `__ldg` free function
+
+This extension adds a single templated free function which may be called from device
+code. This function is not available in host code.
+
+```
+namespace sycl::ext::oneapi::experimental::cuda {
+
+template<typename T>
+T __ldg(const T* ptr);
+
+} // namespace sycl::ext::oneapi::experimental::cuda
+```
+
+`__ldg` returns the data of type `T` located at address `ptr`. The data is cached in the read-only texture cache.
+
+The template parameter `T` can be one of `float`, `vec<float, 2>`, `vec<float, 4>`, `double`, or `vec<double, 2>`.
+
+=== Example of usage
+
+```
+    h.parallel_for<class kernel_name>(range, [=](sycl::nd_item<1> item) {
+      const int idx = item.get_global_id(0);
+      auto cachedA = sycl::ext::oneapi::experimental::cuda::__ldg(&A[idx]);
+      auto cachedB = sycl::ext::oneapi::experimental::cuda::__ldg(&B[idx]);
+      C[idx] = cachedA + cachedB;
+    });
+    });
+```
+
+IMPORTANT: Data returned from `$$__$$ldg`, e.g. `cacheA` and `cacheB` in the above example, must not be written to at any point within the kernel. If such data is written to at any point in the kernel, the code will compile and execute, however the texture cache will not be used. Sometimes the compiler may not be sure that the read-only condition is satisfied. In such cases users can mark the declaration of the pointer used as the argument to `$$__$$ldg` with both the `const` and `$$__$$restrict$$__$$` qualifiers, which may aid the compiler in detecting the read-only condition.
```

Comment:
This is a non-intuitive restriction.  Are you saying that code like this will not work as expected?

```
int v = sycl::ext::oneapi::experimental::cuda::__ldg(&A[idx]);
v = v + 1;
C[idx] = v;  // Value stored might not reflect the increment above
```

Reply:
That code does work as expected, however the texture cache won't be used. Thanks for the comment I see that my wording wasn't exactly correct on this; I've updated it now, hopefully it is clearer? 

`__ldg` here has the same behavior as `__ldg` in CUDA runtime : https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#ldg-function and the clang compiler for nvptx : https://github.com/intel/llvm/blob/6485ec461da8a7abfbd72b930204bb782a1ce3ca/clang/lib/Headers/__clang_cuda_intrinsics.h#L264

I'm just exposing the same functionality to the DPC++ ext_oneapi_cuda backend.

Reply:
Thanks, the new wording makes it clearer.